raft/src/main/java/org/apache/kafka/raft/QuorumState.java

```diff
@@ -336,40 +346,54 @@ public void transitionToUnattached(int epoch) {
      */
     public void transitionToVoted(
         int epoch,
-        int candidateId
+        ReplicaKey candidateKey
     ) {
-        if (localId.isPresent() && candidateId == localId.getAsInt()) {
-            throw new IllegalStateException("Cannot transition to Voted with votedId=" + candidateId +
-                " and epoch=" + epoch + " since it matches the local broker.id");
-        } else if (isObserver()) {
-            throw new IllegalStateException("Cannot transition to Voted with votedId=" + candidateId +
-                " and epoch=" + epoch + " since the local broker.id=" + localId + " is not a voter");
-        } else if (!isVoter(candidateId)) {
-            throw new IllegalStateException("Cannot transition to Voted with voterId=" + candidateId +
-                " and epoch=" + epoch + " since it is not one of the voters " + voters);
-        }
```

Comment:
In KIP-853, replicas that think they are observer need to be allowed to vote if the leader thinks they are voters. Similarly, replicas need to be allowed to vote for replicas even if they think the candidate is not a voter.

In short, only the local set of voters is consider when transitioning to the candidate state.